e2e_tests/tests/api_utils.py

```diff
@@ -192,11 +193,11 @@ def decorator(f: F) -> F:
 # Queries the determined master for resource pool information to determine if agent is used
 # Currently we are assuming that all resource pools are of the same scheduler type
 # which is why only the first resource pool's type is checked.
-@functools.lru_cache(maxsize=1)
-def _get_scheduler_type() -> Optional[bindings.v1SchedulerType]:
-    scheduler_type: Optional[bindings.v1SchedulerType]
+@functools.lru_cache()
```

Comment:
It feels slightly better from an interface perspective to put the optional session parameter (and its creation) into the callers of these primitives (like `skipif_not_hpc`).

Let's take `_get_scheduler_type` and `skipif_not_hpc` explicitly.

If we move the session into `skipif_not_hpc`, then in the case no session is passed (like during all the `skipifs`) then a session will be created inside `skipif_not_hpc`. This session creation involves a login, which may fail after trying to find master if master isn't responsive. So we put that into a try and probably cache `skipif_not_hpc`, too, so we don't mindlessly try to log in a thousand times.

But the functools cache treats each parameter a function gets as necessary, and `skipif_not_hpc` also takes a `reason`. It seems wrong to bust the cache if that reason changes.

I think the design that makes the most sense is for `skipif_not_hpc` to take a required session. But then this makes pytest have to create those sessions explicitly while it's gathering tests. Which, it's at runtime that we'll be most certain of where master has been brought up, so this makes sense. But it's also a bigger refactor than I want to do, here.

In the end, my preference is for this suboptimal solution rather than the solution that moves session creation to callers.

But maybe that's wrong or maybe there's another approach entirely. Thoughts?

Reply:
OOB discussion led to not changing these methods to take an optional session.